src/Api/referrals.js

```diff
@@ -342,34 +317,45 @@ export async function getReferralCodeOwner(chainId, referralCode) {
   const referralStorageAddress = getContract(chainId, "ReferralStorage");
   const provider = getProvider(null, chainId);
   const contract = new ethers.Contract(referralStorageAddress, ReferralStorage.abi, provider);
-  const codeOwner = await contract.codeOwners(referralCode);
-  return codeOwner;
+  try {
+    const codeOwner = await contract.codeOwners(referralCode);
+    return codeOwner;
+  } catch (error) {
+    console.error(error);
+  }
 }
 
 export function useUserReferralCode(library, chainId, account) {
+  const localStorageCode = window.localStorage.getItem(REFERRAL_CODE_KEY);
   const referralStorageAddress = getContract(chainId, "ReferralStorage");
-
-  let { data: userReferralCode, mutate: mutateUserReferralCode } = useSWR(
-    account && [`ReferralStorage:traderReferralCodes`, chainId, referralStorageAddress, "traderReferralCodes", account],
-    {
-      fetcher: fetcher(library, ReferralStorage),
-    }
+  const { data: onChainCode } = useSWR(
+    account && ["ReferralStorage", chainId, referralStorageAddress, "traderReferralCodes", account],
+    { fetcher: fetcher(library, ReferralStorage) }
   );
-
-  const userReferralCodeInLocalStorage = window.localStorage.getItem(REFERRAL_CODE_KEY);
-  if (isHashZero(userReferralCode) && userReferralCodeInLocalStorage) {
-    userReferralCode = userReferralCodeInLocalStorage;
-  }
-
-  let userReferralCodeString;
-  if (userReferralCode && !isHashZero(userReferralCode)) {
-    userReferralCodeString = decodeReferralCode(userReferralCode);
-  }
+  const { data: localStorageCodeOwner } = useSWR(
+    localStorageCode &&
+      REGEX_VERIFY_BYTES32.test(localStorageCode) && [
+        "ReferralStorage",
+        chainId,
+        referralStorageAddress,
+        "codeOwners",
+        localStorageCode,
+      ],
+    { fetcher: fetcher(library, ReferralStorage) }
+  );
+  const [isCodeAttached, userReferralCode, userReferralCodeString] = useMemo(() => {
+    if (onChainCode && !isHashZero(onChainCode)) {
+      return [true, onChainCode, decodeReferralCode(onChainCode)];
+    } else if (localStorageCodeOwner && !isAddressZero(localStorageCodeOwner)) {
+      return [false, localStorageCode, decodeReferralCode(localStorageCode)];
+    }
+    return [ethers.utils.AddressZero];
+  }, [localStorageCode, localStorageCodeOwner, onChainCode]);
 
   return {
     userReferralCode,
     userReferralCodeString,
-    mutateUserReferralCode,
+    isCodeAttached,
```

Comment:
maybe `savedOnChain`? i don't want to bother with naming however for new developers without knowing the context "is attached" might be unclear when they see it in some component's code, makes sense?

Reply:
Yeah sure `attachedOnChain` makes sense.